File: solutions/kotlin/run-length-encoding/1/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {

    fun encode(input: String): String {
        if (input.isEmpty()) return ""
        val first = input.first()
        val count = input.takeWhile { it == first }.count()

        return when {
            count > 1 -> "$count"
            else -> ""
        } + first + encode(input.substring(count))
    }

    fun decode(input: String): String {
        if (input.isEmpty()) return ""
        val digits = input.takeWhile { it.isDigit() }
        val tail = input.substring(digits.count())
        val repeated = tail.first().toString().repeat(digits.toIntOrNull() ?: 1)
        return repeated + decode(tail.substring(1))
    }
}

File: solutions/csharp/run-length-encoding/1/RunLengthEncoding.cs
using System;
using System.Linq;

public static class RunLengthEncoding
{
    public static string Encode(string input)
    {
        if (string.IsNullOrEmpty(input)) return "";
        var first = input.First();
        var parts = input.TakeWhile(x => x == first).ToList();
        if (parts.Count == 1) return $"{first}" + Encode(input.Substring(parts.Count));
        return $"{parts.Count}{first}" + Encode(input.Substring(parts.Count));
    }
    
    public static string Decode(string input)
    {
        if (string.IsNullOrEmpty(input)) return "";
        var digits = input.TakeWhile(x => Char.IsDigit(x)).ToList();
        if (digits.Any())
        {
            var count = int.Parse(string.Join("", digits));
            var rest = input.Substring(digits.Count());
            var letter = rest.First();
            var decode = string.Join("", Enumerable.Repeat(letter, count));
            return decode + Decode(rest.SafeSubstring(1));
        } 
        else 
        {
            var first = input.First().ToString();
            return first + Decode(input.SafeSubstring(1));
        }
    }
    
    public static string SafeSubstring(this string str, int startIndex)
    {
        if (str.Length > startIndex) return str.Substring(startIndex);
        return "";
    }
}
